oxidation/libparsec/crates/libparsec_fs/src/file_operations.rs

```diff
@@ -0,0 +1,358 @@
+        let substop = min(offset + size, blockstart + blocksize);
+        let subsize = substop
+            .checked_sub(substart)
+            .expect("Substop is always greater than substart");
+        let content_offset = substart - offset;
+
+        // Prepare new chunk
+        let new_chunk = Chunk::new(
+            substart,
+            NonZeroU64::new(substart + subsize)
+                .expect("subsize is always strictly greater than zero"),
+        );
+        write_operations.push((new_chunk.clone(), content_offset as i64 - padding as i64));
+
+        // Get the corresponding chunks
+        let new_chunks = match manifest.get_chunks(block as usize) {
+            Some(block_chunks) => {
+                let (new_chunks, more_removed_ids) =
+                    block_write(block_chunks, subsize, substart, new_chunk);
+                removed_ids.extend(more_removed_ids);
+                new_chunks
+            }
+            None => vec![new_chunk],
+        };
+
+        // Update data structures
+        if blocks.len() == block as usize {
+            blocks.push(new_chunks);
+        } else {
+            blocks[block as usize] = new_chunks;
+        }
+    }
+
+    // Evolve manifest
+    let new_size = max(manifest.size, offset + size);
+    let new_manifest = manifest
+        .clone()
+        .evolve_and_mark_updated(new_size, blocks, timestamp);
+
+    (new_manifest, write_operations, removed_ids)
+}
+
+// Prepare truncate
+
+pub fn prepare_truncate(
+    manifest: &LocalFileManifest,
+    size: u64,
+    timestamp: DateTime,
+) -> (LocalFileManifest, HashSet<ChunkID>) {
+    // Check that there is something to truncate
+    if size >= manifest.size {
+        return (manifest.clone(), HashSet::new());
+    }
+
+    // Find limit block
+    let blocksize = u64::from(manifest.blocksize);
+    let block = size / blocksize;
+    let remainder = size % blocksize;
+
+    // Prepare removed ids and new blocks
+    let mut removed_ids: HashSet<ChunkID> = manifest
+        .blocks
+        .get(block as usize..)
+        .unwrap_or_default()
+        .iter()
+        .flatten()
+        .map(|x| x.id)
+        .collect();
+    let mut new_blocks = manifest
+        .blocks
+        .get(0..block as usize)
+        .unwrap_or_default()
+        .to_vec();
+
+    // Last block needs to be split
+    if remainder != 0 {
+        let chunks = manifest
+            .get_chunks(block as usize)
+            .expect("Block is expected to be part of the manifest");
+
+        // Find the index of the last chunk to include
+        let chunk_index = match chunks.binary_search_by_key(&size, |x| x.start) {
+            Ok(x) => x - 1,
+            Err(x) => x - 1,
+        };
+
+        // Create the new last chunk
+        let last_chunk = chunks
+            .get(chunk_index)
+            .expect("The index is found using binary search and hence always valid");
+        let mut new_chunk = last_chunk.clone();
```

Comment:
last_chunk or new_chunk ? We should try to avoid unecessary redirection

Reply:
The new chunk is a modified version of the last chunk, is there a problem with that? What do you mean by "redirection"?

Reply:
One of these variable is unecessary (can be grouped):
```Rust
chunks
            .get(chunk_index)
            .expect("The index is found using binary search and hence always valid")
            .clone();
```

Reply:
Oh I see. Isn't this kind of pattern optimized away by the compiler? I'd rather not think about that while reading or writing code and focus on the meaning. Here, `last_chunk` is the last chunk to consider in the original manifest and `new_chunk` is the new one we're inserting in the new manifest, so IMO the code is more readable as it is.